File: amic-platform/src/modules/kiis/types/deal.ts
/** 딜 아이템 (GET /deals/by-company/{corpCode} → items[]) */
export interface DealItem {
  id: number;
  company_id: number | null;
  investor_name: string | null;
  target_company: string;
  target_company_id: number | null;
  amount: string | null;
  amount_display: string | null;
  round_stage: string | null;
  sector: string | null;
  deal_date: string | null;
  deal_year: number | null;
  source_url: string | null;
  source_type: string | null;
  is_lead_investor: boolean;
  fund_id: number | null;
  fund_code: string | null;
  fund_name: string | null;
}

/** 딜 목록 아이템 (축약) */
export interface DealListItem {
  id: number;
  target_company: string;
  amount_display: string | null;
  round_stage: string | null;
  sector: string | null;
  deal_date: string | null;
}

/** 딜 목록 래퍼 응답 */
export interface DealListResponse {
  total: number;
  page: number;
  size: number;
  items: DealItem[];
}

/** 섹터별 집계 */
export interface SectorAggregation {
  sector: string;
  sector_name: string;
  deal_count: number;
  total_amount: string | null;
}

/** 섹터별 집계 래퍼 응답 */
export interface SectorAggregationResponse {
  total_deals: number;
  items: SectorAggregation[];
}

/** 스테이지별 집계 */
export interface StageAggregation {
  stage: string;
  stage_name: string;
  deal_count: number;
  total_amount: string | null;
}

/** 스테이지별 집계 래퍼 응답 */
export interface StageAggregationResponse {
  total_deals: number;
  items: StageAggregation[];
}

/** 연도별 트렌드 */
export interface YearlyTrend {
  year: number;
  deal_count: number;
  total_amount: string | null;
}

/** 트렌드 래퍼 응답 */
export interface TrendResponse {
  corp_code: string | null;
  corp_name: string | null;
  items: YearlyTrend[];
}

/** Deal 집계 공통 파라미터 (by-sector, by-stage) */
export interface DealAggregationParams {
  year?: number;
  years?: number;
  corp_code?: string;
}

/** Deal 트렌드 파라미터 (GET /deals/trends — year 미지원) */
export interface DealTrendParams {
  years?: number;
  corp_code?: string;
}

/** Deal by-company 조회 파라미터 */
export interface DealByCompanyParams {
  years?: number;
  sector?: string;
  stage?: string;
  page?: number;
  size?: number;
}

/** Deal by-fund 조회 파라미터 */
export interface DealByFundParams {
  years?: number;
  page?: number;
  size?: number;
}

/** 투자 규모 분포 구간 */
export interface AmountBucket {
  bucket_label: string;
  bucket_min: number;
  bucket_max: number | null;
  deal_count: number;
  total_amount: string | null;
}

/** 투자 규모 통계 (GET /deals/stats) */
export interface DealAmountStats {
  total_deals: number;
  total_amount: string | null;
  avg_amount: string | null;
  median_amount: string | null;
  min_amount: string | null;
  max_amount: string | null;
  distribution: AmountBucket[];
}

/* ─── 투자성향 정성적 요약 (GET /deals/tendency-summary) ─── */

/** 투자성향 - 대표 딜 아이템 */
export interface TendencyDealItem {
  target_company: string;
  amount_display: string | null;
  round_stage: string | null;
  deal_date: string | null;
  source_url: string | null;
}

/** 투자성향 - 섹터별 상세 */
export interface TendencySectorDetail {
  sector: string;
  sector_name: string;
  deal_count: number;
  total_amount: string | null;
  total_amount_display: string | null;
  percentage: number;
  description: string;
  deals: TendencyDealItem[];
}

/** 투자성향 - 스테이지별 상세 */
export interface TendencyStageDetail {
  stage: string;
  stage_name: string;
  deal_count: number;
  total_amount: string | null;
  total_amount_display: string | null;
  percentage: number;
  description: string;
  deals: TendencyDealItem[];
}

/** 투자성향 정성적 요약 응답 */
export interface TendencySummaryResponse {
  corp_code: string;
  years: number;
  total_deals: number;
  total_amount: string | null;
  total_amount_display: string | null;
  summary_text: string;
  sector_summary: string;
  stage_summary: string;
  sectors: TendencySectorDetail[];
  stages: TendencyStageDetail[];
}

File: kiis/app/schemas/deal.py
"""딜 소싱 스키마"""

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class DealItem(BaseModel):
    """딜 상세 정보"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    company_id: int | None = Field(None, description="투자사 ID")
    investor_name: str | None = Field(None, description="투자사명")
    target_company: str = Field(..., description="피투자사명")
    target_company_id: int | None = Field(None, description="피투자사 ID")
    amount: Decimal | None = Field(None, description="투자 금액 (원)")
    amount_display: str | None = Field(None, description="투자 금액 표시용")
    round_stage: str | None = Field(None, description="투자 단계")
    sector: str | None = Field(None, description="투자 섹터")
    deal_date: date | None = Field(None, description="거래일")
    deal_year: int | None = Field(None, description="거래 연도")
    source_url: str | None = Field(None, description="출처 URL")
    source_type: str | None = Field(None, description="출처 유형")
    is_lead_investor: bool = Field(False, description="리드 투자사 여부")
    fund_id: int | None = Field(None, description="펀드 ID")
    fund_code: str | None = Field(None, description="펀드 코드")
    fund_name: str | None = Field(None, description="펀드명")


class DealListItem(BaseModel):
    """딜 목록 아이템 (간략)"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    target_company: str
    amount_display: str | None
    round_stage: str | None
    sector: str | None
    deal_date: date | None


class DealListResponse(BaseModel):
    """딜 목록 응답"""

    total: int = Field(..., description="총 건수")
    page: int = Field(..., description="페이지 번호")
    size: int = Field(..., description="페이지 크기")
    items: list[DealItem] = Field(default_factory=list, description="딜 목록")


class SectorAggregation(BaseModel):
    """섹터별 집계"""

    sector: str = Field(..., description="섹터 코드")
    sector_name: str = Field(..., description="섹터 표시명")
    deal_count: int = Field(..., description="딜 수")
    total_amount: Decimal | None = Field(None, description="총 투자 금액 (원)")


class SectorAggregationResponse(BaseModel):
    """섹터별 집계 응답"""

    total_deals: int = Field(..., description="총 딜 수")
    items: list[SectorAggregation] = Field(default_factory=list, description="섹터별 집계")


class StageAggregation(BaseModel):
    """단계별 집계"""

    stage: str = Field(..., description="투자 단계 코드")
    stage_name: str = Field(..., description="단계 표시명")
    deal_count: int = Field(..., description="딜 수")
    total_amount: Decimal | None = Field(None, description="총 투자 금액 (원)")


class StageAggregationResponse(BaseModel):
    """단계별 집계 응답"""

    total_deals: int = Field(..., description="총 딜 수")
    items: list[StageAggregation] = Field(default_factory=list, description="단계별 집계")


class YearlyTrend(BaseModel):
    """연도별 트렌드"""

    year: int = Field(..., description="연도")
    deal_count: int = Field(..., description="딜 수")
    total_amount: Decimal | None = Field(None, description="총 투자 금액 (원)")


class TrendResponse(BaseModel):
    """트렌드 응답"""

    corp_code: str | None = Field(None, description="투자사 DART 고유번호")
    corp_name: str | None = Field(None, description="투자사명")
    items: list[YearlyTrend] = Field(default_factory=list, description="연도별 트렌드")


class AmountBucket(BaseModel):
    """투자 금액 구간별 분포"""

    bucket_label: str = Field(..., description="구간 라벨 (예: '10억 미만')")
    bucket_min: int = Field(..., description="구간 하한 (억원)")
    bucket_max: int | None = Field(None, description="구간 상한 (억원)")
    deal_count: int = Field(..., description="해당 구간 딜 수")
    total_amount: Decimal | None = Field(None, description="해당 구간 합계 (원)")


class DealAmountStats(BaseModel):
    """투자 규모 통계"""

    total_deals: int = Field(..., description="총 딜 수")
    total_amount: Decimal | None = Field(None, description="총 투자 금액 (원)")
    avg_amount: Decimal | None = Field(None, description="평균 투자 금액 (원)")
    median_amount: Decimal | None = Field(None, description="중앙값 (원)")
    min_amount: Decimal | None = Field(None, description="최소 투자 금액 (원)")
    max_amount: Decimal | None = Field(None, description="최대 투자 금액 (원)")
    distribution: list[AmountBucket] = Field(
        default_factory=list, description="금액 구간별 분포"
    )


class DealCreateRequest(BaseModel):
    """딜 생성 요청"""

    target_company: str = Field(..., min_length=1, description="피투자사명")
    amount: Decimal | None = Field(None, description="투자 금액 (원)")
    amount_display: str | None = Field(None, description="투자 금액 표시용")
    round_stage: str | None = Field(None, description="투자 단계")
    sector: str | None = Field(None, description="투자 섹터")
    deal_date: date | None = Field(None, description="거래일")
    source_url: str | None = Field(None, description="출처 URL")
    source_type: str | None = Field(None, description="출처 유형")
    is_lead_investor: bool = Field(False, description="리드 투자사 여부")
    description: str | None = Field(None, description="딜 설명")


class TendencyDealItem(BaseModel):
    """투자성향 - 대표 딜 아이템"""

    target_company: str
    amount_display: str | None = None
    round_stage: str | None = None
    deal_date: date | None = None
    source_url: str | None = None


class TendencySectorDetail(BaseModel):
    """투자성향 - 섹터별 상세"""

    sector: str
    sector_name: str
    deal_count: int
    total_amount: Decimal | None = None
    total_amount_display: str | None = None
    percentage: float
    description: str
    deals: list[TendencyDealItem] = Field(default_factory=list)


class TendencyStageDetail(BaseModel):
    """투자성향 - 스테이지별 상세"""

    stage: str
    stage_name: str
    deal_count: int
    total_amount: Decimal | None = None
    total_amount_display: str | None = None
    percentage: float
    description: str
    deals: list[TendencyDealItem] = Field(default_factory=list)


class TendencySummaryResponse(BaseModel):
    """투자성향 정성적 요약 응답"""

    corp_code: str
    years: int
    total_deals: int
    total_amount: Decimal | None = None
    total_amount_display: str | None = None
    summary_text: str
    sector_summary: str
    stage_summary: str
    sectors: list[TendencySectorDetail] = Field(default_factory=list)
    stages: list[TendencyStageDetail] = Field(default_factory=list)


class DealExtractRequest(BaseModel):
    """뉴스에서 딜 추출 요청"""

    news_article_id: int = Field(..., description="뉴스 기사 ID")
    investor_corp_code: str | None = Field(None, description="투자사 DART 고유번호 (선택)")
    fund_code: str | None = Field(None, description="펀드 코드 (펀드 단위 딜 추적)")
